runtime/protos/runtime.proto

```diff
@@ -17,232 +17,270 @@ option go_package = "github.com/ServiceWeaver/weaver/runtime/protos";
-  string error = 11;
+// HealthStatus specifies the health of a weavelet.
+enum HealthStatus {
+  UNKNOWN = 0;
+  HEALTHY = 1;
+  UNHEALTHY = 2;
+  TERMINATED = 3;
+}
 
-  // Request to get the address for a listener.
-  GetAddressRequest get_address_request = 14;
+// GetMetricsRequest is a request from an envelope for a weavelet's metrics.
+message GetMetricsRequest {}
 
-  // Request to export a listener.
-  ExportListenerRequest export_listener_request = 15;
+// GetMetricsReply is a reply to a GetMetricsRequest.
+//
+// Note that when a weavelet receives a GetMetricsRequest, it does not reply
+// with a snapshot of every metric. Doing so would be wasteful as some metrics
+// may not have changed since the previous GetMetricsRequest. Instead, the
+// weavelet responds only with the values of the metrics that changed since the
+// previous GetMetricsRequest.
+message GetMetricsReply {
+  MetricUpdate update = 1;
 }
 
-// Deployment holds internal information necessary for an application
-// deployment.
-message Deployment {
-  string id = 1;  // Unique identifier for the deployment.
-
-  // Application configuration.
-  AppConfig app = 2;
+// MetricType specifies a particular kind of metric.
+enum MetricType {
+  INVALID = 0;
+  COUNTER = 1;
+  GAUGE = 2;
+  HISTOGRAM = 3;
+}
 
-  // Run as a single process locally.
-  bool single_process = 10;
+// MetricUpdate contains metric updates.
+message MetricUpdate {
+  // Metric definitions and updated values.
+  //
+  // All metrics are identified by their globally unique id. Two endpoints
+  // exchanging metric updates must ensure that a received metric value has been
+  // preceded by a metric definition with the same id, either in the current
+  // update or an earlier one.
+  repeated MetricDef defs = 1;      // metric definitions
+  repeated MetricValue values = 2;  // metric values
 }
 
-// A group of components.
-message ComponentGroup {
-  repeated string components = 1;
+// MetricDef is the definition of a new metric.
+message MetricDef {
+  uint64 id = 1;                   // metric's unique id
+  string name = 2;                 // name of the metric
+  MetricType typ = 3;              // type of metric
+  string help = 4;                 // metric's help message
+  map<string, string> labels = 5;  // metric labels
+  repeated double bounds = 6;      // histogram bucket bounds
 }
 
-// AppConfig holds configuration for a Service Weaver application.
-message AppConfig {
-  // Application name (derived from binary if missing).
-  string name = 2;
+// MetricValue is the value associated with a metric.
+message MetricValue {
+  uint64 id = 1;               // metric's unique id.
+  double value = 2;            // value for counter and gauge, sum for histogram
+  repeated uint64 counts = 3;  // histogram counts
+}
 
-  // File name of the binary that implements the app.
-  string binary = 1;
+// MetricSnapshot is a snapshot of a metric. It is the union of a MetricDef and
```

Comment:
Note: this is also not used in the protocol, and maybe belongs in a separate file?